deploy/buildtools/buildfarm.py

```diff
@@ -0,0 +1,368 @@
+        Args:
+            args: Args (i.e. options) passed to the build farm.
+        """
+        self.args = args
+        self.build_hosts = []
+
+    @staticmethod
+    @abc.abstractmethod
+    def NAME() -> str:
+        """Human-readable name (used as the "build-farm-type" in the YAML).
+
+        Returns:
+            Human-readable name.
+        """
+        return ""
+
+    @abc.abstractmethod
+    def request_build_host(self, build_config: BuildConfig) -> None:
+        """Request build host to use for build config.
+
+        Args:
+            build_config: Build config to request build host for.
+        """
+        return
+
+    @abc.abstractmethod
+    def wait_on_build_host_initialization(self, build_config: BuildConfig) -> None:
+        """Ensure build host is launched and ready to be used.
+
+        Args:
+            build_config: Build config used to find build host that must ready.
+        """
+        return
+
+    def get_build_host(self, build_config: BuildConfig) -> BuildHost:
+        """Get build host associated with the build config.
+
+        Args:
+            build_config: Build config used to find the build host.
+
+        Returns:
+            Build host associated with the build config.
+        """
+        for build_host in self.build_hosts:
+            if build_host.build_config == build_config:
+                return build_host
+
+        raise Exception(f"Unable to find build host for {build_config.name}")
+
+    def get_build_host_ip(self, build_config: BuildConfig) -> str:
+        """Get IP address associated with this dispatched build host.
+
+        Args:
+            build_config: Build config to find build host for.
+
+        Returns:
+            IP address for the specific build host.
+        """
+        build_host = self.get_build_host(build_config)
+        ip_address = build_host.ip_address
+        assert ip_address is not None, f"Unassigned IP address for build host: {build_host}"
+        return ip_address
+
+    @abc.abstractmethod
+    def release_build_host(self, build_config: BuildConfig) -> None:
+        """Release the build host.
+
+        Args:
+            build_config: Build config to find build host to terminate.
+        """
+        return
+
+
+class IPAddrBuildFarm(BuildFarm):
+    """Build farm that selects from a set of user-determined IPs to allocate a new build host.
+
+    Attributes:
+        build_hosts_allocated: Count of build hosts assigned with builds (`BuildConfig`s).
+    """
+    build_hosts_allocated: int
+
+    def __init__(self, args: Dict[str, Any]) -> None:
+        """
+        Args:
+            args: Args (i.e. options) passed to the build farm.
+        """
+        super().__init__(args)
+
+        self._parse_args()
+
+    @staticmethod
+    def NAME() -> str:
+        """Human-readable name (used as the "build-farm-type" in the YAML).
+
+        Returns:
+            Human-readable name.
+        """
+        return "unmanaged"
```

Comment:
Its not nice to have the users put the class name into the .yaml file. Instead, we added this human-readable name that you can refer to in the .yaml (instead of writing EC2BuildFarm).

Reply:
Why? It's adding an additional level of indirection for no reason. If you just use the real name, users can easily find the corresponding class definition, instead of searching the codebase for "unmanaged".

Even apart from that, why are the names are completely different?: IPAddrBuildFarm vs. unmanaged.

Reply:
I thought we talked about this in our meetings and i'm sure there's something buried in the PRs for where-to-build. Both @timsnyder-siv and I thought this was preferable, because the class name doesn't really line up with what you want to put here (the class name should probably be more verbose to disambiguate itself in the namespace/describe what it is, where as the key is inherently scoped). It also makes the name of the class itself part of the user API making it hard to rename internally. 

I'm down to rename it from unmanaged if there's something you prefer.

Reply:
I was under the impression that we came to the opposite conclusion when I brought this up in the meeting. In any case, I'm going to punt this to #1010 and resolve this for this PR.